### DISCUSSION TITLE
[macOS] [Older Support] [macOS 10.8 Mountain Lion] DevilutionX doesn't open

### DISCUSSION THREAD

**Discussion by shinra-electric** [diasurgical/DevilutionX#2861] [Q&A]
**Important information**
macOS 10.8.5 Mountain Lion
DevilutionX 1.2.1 release (Also tried many older releases)

**Describe**
DevilutionX does not open. There is a small bump in the dock but it disappears almost immediately

**To Reproduce**
Try to open DevilutionX

**Expected behavior**
Expect the app to open.

**Additional context**
I looked for crash logs, but I couldn't find any. 
I realise that this is an old OS, but I saw other issues relating to PPC support which is even older...

**Comment by AJenbo** [comment]
DevilutionX releases for macOS is build on High Sierra (10.13), which I believe limits the support to 10.11+. For older versions of macOS you will need to [build from source](https://github.com/diasurgical/devilutionX/blob/master/docs/building.md).
Note that Homebew only has officially full support for 10.14+, and partial support for 10.09+, there is [a fork for Mac OS 10.4 - 10.6](https://github.com/mistydemeo/tigerbrew) which leaves you in a bit of a predicament, but maybe one of them will still work. If not you will have to figure out how to get the needed dependencies (cmake, and sdl2_ttf) on your system manually.
Let us know how it goes and feel free to ask for additional help on the chat.

Regarding support for PPC; Mac OS is not the only operating system that run/ran on PowerPC, so most of that is about Linux or Amiga :)

**Reply by shinra-electric** [reply]
Hmm, I'll give it a shot. I have DevilutionX on a modern Mac, so I'm only doing this for a bit of fun. 

Can you confirm about the macOS build though? I would have thought that it's the Xcode version that limits the build targets, rather than the actual OS that it's built on. (From what I understand cmake points to Xcode and uses its compiler). Which version of Xcode dropped support for 10.8? 

(I know that Xcode 10 dropped support for compiling 32-bit apps, that may or may not be relevant).

**Reply by AJenbo** [reply]
By macOS 10.13 I'm implying XCode 10.1. You will probably need XCode 5.1.1.

**Reply by shinra-electric** [reply]
Ok, I'll start by trying to build the same as the current release, and if I can get it to work I'll try to go backwards. I'm downloading previous versions of Xcode now, and will set up some VMs to test them out. 

(I'm not going to try to build on Mountain Lion, I'm sure later versions of Xcode can target it. 5.1.1 might be a bit old and painful to use).

**Reply by AJenbo** [reply]
Did you manage to get things running?

**Reply by shinra-electric** [reply]
Em, no… I got distracted by other things. 😅

I’ll give it a try later this week if I get the time.

**Reply by AJenbo** [reply]
Ok, things have gotten a bit simpler since last. We now have less external dependencies, and there is also a tarball for 1.3.0 with all dependencies included in case that tickles you fancy (helpful if you can't establish a HTTPS connection to GitHub) :)

**Reply by shinra-electric** [reply]
So I had a first shot at this.... 

First I tried to build from the tarball using my M1 using the build instructions, and it worked fine. 

Then I installed High Sierra 10.13 on my Intel MacBook Air, and after a bit of a struggle it built and ran too. 

Since I want to avoid building on Mountain Lion (Homebrew is not supported there) I generated an Xcode project using `cmake -G Xcode -H. -B_build`  

The Xcode project opened up fine. I set the target to 10.13 first to just make sure it's working, but it failed with a missing file in the libzt Debug folder. So I'm sure I'm not setting it to build for release correctly... 

![Screen Shot 2021-11-30 at 23 09 46](https://user-images.githubusercontent.com/50119606/144145418-7bde53ca-e532-4d98-9ec8-25fb2100c3e8.png)

If I can get Xcode to build it fine, my plan is just to set the target OS back further to 10.8. 

If you have any advice, it would be appreciated. My experience with Xcode is limited to making SwiftUI apps...

**Reply by AJenbo** [reply]
Interesting considering I build the release on Mac OS 10.13.6 :)

To skip some dependencies you can run `cmake -G Xcode -H. -B_build -DNONET=ON -DNOSOUND=ON`, that should at least get you past the `libzt` issue.

I'm running Xcode 10.1. Targeting 10.9 seems to work, but if I set it to 10.8 it trigger a couple of errors:
![image](https://user-images.githubusercontent.com/204594/144147405-5980616e-9c6f-4a26-b3e4-ddb079859f03.png)
They do seam to be a bit of a show stopper if these headers are truly missing from the SDK :/

You can also set the target version without using Xcode via this line: https://github.com/diasurgical/devilutionX/blob/master/CMakeLists.txt#L1138

> My experience with Xcode is limited to making SwiftUI apps...

Speaking of Xcode UI stuff, if you know how to add a Launch Screen to an iOS app that would be greatly appreciated here: https://github.com/diasurgical/devilutionX/pull/3612 :)

**Reply by shinra-electric** [reply]
I wasn't able to download a branch which that pull request and generate the Xcode project file, but that's just due to my lack of GitHub ability. I'm not sure how you are creating the iOS project, but [here](https://betterprogramming.pub/launch-screen-with-swiftui-bd2958771f3b) is a quick guide for making the splash screen using SwiftUI. It's just setting things in the info.plist. (Note that it is probably different for UIKit...). This guide is using Xcode 12 or later, I have no idea how to do it on earlier Xcode versions...

Regarding the errors building for 10.8, while I cannot claim to fully understand I think that the issue is due to libc++ or libstdc++. I found [this](https://trac.macports.org/wiki/FAQ#libcpp) on the MacPorts FAQ page. Those "unfound header files" are actually from the libc++ standard library, but for 10.8 it is being build with libstdc++ so they are not found.

**Reply by shinra-electric** [reply]
Oh, and speaking of MacPorts: Homebrew has only official support for 10.15 and later (with unofficial support for 10.10 and later), but MacPorts has support [for all OSs](https://www.macports.org/install.php#installing) going back to Tiger...

**Reply by AJenbo** [reply]
If you can sort it out I´ll be happy to merge various fixes for supporting older versions of Mac OS X :)

**Reply by shinra-electric** [reply]
So, I've made a bit of progress. Just by changing the C++ Standard Library setting to use libc++ it will now compile for 10.8. 

![Screen Shot 2021-12-09 at 15 06 27](https://user-images.githubusercontent.com/50119606/145422235-044ece85-defc-4baf-8a19-c50e998fef70.png)
![Screen Shot 2021-12-09 at 15 07 03](https://user-images.githubusercontent.com/50119606/145422241-496c69ee-080c-480e-9386-81fd12d32b86.png)

However, when trying to archive, I'm getting a no-file-found error for the png dep.

![Screen Shot 2021-12-09 at 15 07 38](https://user-images.githubusercontent.com/50119606/145422499-8de3ce61-8059-45d0-b8c9-0eba5117122d.png)

Would you have any insight into this?

**Reply by AJenbo** [reply]
I normally have to do that with sudo enabled (i usually do this via a terminal).

I'm guessing this is the part it is having issues with: https://github.com/diasurgical/devilutionX/blob/master/CMakeLists.txt#L979

**Reply by shinra-electric** [reply]
I had another shot at this today. Got a little bit further... 

So since Xcode was being funky, I thought I'd try compiling with Cmake instead. So I changed the macOS target to 10.8, and then did the normal build steps but passed in a flag to force Cmake to build with libc++ 

>cmake -S. -Bbuild -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_FLAGS="-stdlib=libc++"

Everything built fine, and I ended up with a working binary (on High Sierra and Monterey), but there were warnings that libsodium was built for a higher OS than the target (10.8), and sure enough the binary didn't work on Mountain Lion. 

From what I read about libsodium, it handles the signing and whatnot, so it might explain why the binary didn't run on ML.

**Reply by AJenbo** [reply]
libsodium is used for network encryption for password protected games. I suggest you apply -DNONET=ON to disable network support and avoid ZeroTier, LibSodium and ASIO which are pretty huge and non trivial to port.

**Reply by glebm** [reply]
This is probably because you're using libsodium from homebrew, which doesn't support 10.8?
You can compile it from source by simply setting `-DDEVILUTIONX_SYSTEM_LIBSODIUM=OFF`

**Reply by shinra-electric** [reply]
I followed glebm's advice and built with: 

>cmake -S. -Bbuild -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_FLAGS="-stdlib=libc++" -DDEVILUTIONX_SYSTEM_LIBSODIUM=OFF

Everything went smoothly, no warnings about libsodium, and again the binary worked on High Sierra. 

But again, it crashed on launch on ML. Looking at the crash log it said: 
>Dyld Error Message: 
>   Library not loaded: /usr/local/opt/fmt/lib/libfmt.8.dylb
>   Referenced from: [Path to DevilutionX]
>   Reason: image not found

So I went there, and sure enough it didn't exist. I installed libfmt through Macports (Again? Not sure if I did that before) but Macports installed to a different location: opt/local/lib/ 

I created a directory in /usr/local/opt/fmt/lib/ and threw in an alias for libfmt.8.dylb there, and..... crash on launch with the same error message....

**Reply by glebm** [reply]
@shinra-electric Not sure what's up with that but you can also do the same for libmft `-DDEVILUTIONX_SYSTEM_LIBFMT=OFF`

**Reply by AJenbo** [reply]
Don't forget to include the required liberies or you won't be able to move the application to another computer: `-DMACOSX_STANDALONE_APP_BUNDLE=ON` (this was renamed recently)

**Reply by shinra-electric** [reply]
`-DMACOSX_STANDALONE_APP_BUNDLE=ON` was not recognised. Are you sure about the syntax?

**Reply by AJenbo** [reply]
It was renamed from DIST 4 says ago, so if you haven't updated after that you would have to use the old name.

**Reply by shinra-electric** [reply]
I cloned the repo, added the app bundle command, everything built with no errors. Again, works on HS, crash on launch on ML...

I'm pretty much out of ideas at this stage, might give up soon. 

I was going to try to build on my ML laptop, but the whole GitHub block for TLS 1.0 is preventing me from installing stuff... :(

**Reply by glebm** [reply]
@shinra-electric Generally you have to build on a system at least as old as what you're trying to run it on! If you don't do that, your executable is linked against newer versions of system libraries, including libc, which isn't going to work no matter what flag you pass.

> I was going to try to build on my ML laptop, but the whole GitHub block for TLS 1.0 is preventing me from installing stuff... :(

We have a tool to make a source archive with dependencies. Here is how you can use it.

On your newer Mac:

1. Add `'sdl2', 'sdl_image', 'libpng', 'libfmt', 'bzip2', 'libsodium'` to `tools/make_src_dist.py` here: https://github.com/diasurgical/devilutionX/blob/467ed9dd9da358dbcd19fdf7449e33603933c697/tools/make_src_dist.py#L37
2. Run `tools/make_src_dist.py`. This will produce an archive that includes all of the dependencies.
3. Copy this archive from your new Mac to your old Mac.
4. Build on your old Mac with the following flags:

   ```
   -DDEVILUTIONX_SYSTEM_SDL2=OFF -DDEVILUTIONX_SYSTEM_SDL_IMAGE=OFF -DDEVILUTIONX_SYSTEM_LIBPNG=OFF -DDEVILUTIONX_SYSTEM_LIBFMT=OFF -DDEVILUTION_SYSTEM_BZIP2=OFF -DDDEVILUTION_SYSTEM_LIBSODIUM=OFF
   ```

**Reply by shinra-electric** [reply]
I'll give it a try tomorrow, but there will be errors because ML uses stdlibc++ and not libc++, and it will not recognise the headers....

**Reply by glebm** [reply]
@shinra-electric Tomorrow it should be a bit easier, I've sent a PR that adds a `--fully_vendored` flag to the `tools/make_src_dist.py` command #3742

**Reply by AJenbo** [reply]
I still recommend that you try -DNONET=ON and -DNOSOUND=ON so that you can exclude as many dependencies as possible and rule them out as potential sources for the problem.

**Reply by glebm** [reply]
The new instructions, from the current master, are:

1. Run `python3 tools/make_src_dist.py --fully_vendored` on your new Mac.
2. Copy the resulting `tar.xz` archive to your old Mac.
3. Unpack the archive on your old Mac (`tar xf devilutionx-src-*.tar.xz`) and build.

**Reply by shinra-electric** [reply]
Sorry, I'm having trouble with the python script. It hasn't worked with the same errors on Monterey, Mojave and High Sierra (Intel), or Monterey (M1). The python3 version is 3.9.9 (I have 3.10.0 installed, but 3.9.9 is used for some reason). 
![Screenshot 2021-12-13 at 13 38 57](https://user-images.githubusercontent.com/50119606/145822638-b5540a51-e474-415e-8882-8483d2e718e9.png)

**Reply by AJenbo** [reply]
You need to build and install install smpq first: https://github.com/diasurgical/devilutionX/blob/master/tools/build_and_install_smpq.sh

**Reply by shinra-electric** [reply]
That doesn't work for me either. I have all the declared dependencies installed. 

There are a few errors that mention libTomMalloc and libTomFree, is libTom also a dependency?

**Reply by glebm** [reply]
> There are a few errors

Post the errors here and I may be able to help. AFAIK that library is vendored within StormLib but maybe not?
